File: src/functions/growth_argmaxi.rs
use super::abs;
use super::argmaxi::argmaxi;
use super::expand::expand_at_dim;
use super::expand_for_batches::expand_for_batches;
use super::gather::gather;
use super::mark_reserved_indices::mark_reserved_indices;
use super::move_to_back::move_to_back;
use super::reduce::reduce;
use super::reduce_specialized;
use super::reshape::reshape;
use super::scatter::scatter;
use super::sum::sum_generic;
use super::transpose::transpose_dims;
use super::wheres::{where_op, where_value};
use super::zeros_like::zeros_like;
use super::{abs::abs_ndarray, expand};
use ndarray::{ArrayD, Axis, ScalarOperand};
use ndarray::{NdFloat, Zip};
use num_traits::{Float, NumCast, ToPrimitive, Zero};
use std::fmt::Debug;

/// Finds indices with maximum growth potential based on counts and values
///
/// # Arguments
/// * `x` - Values array of floating point type
/// * `counts` - Count values of integer type
/// * `eps` - Small epsilon value to avoid division by zero
/// * `threshold` - Threshold for considering growth potential
///
/// # Returns
/// * Indices of values with maximum growth potential
pub fn growth_argmaxi<T, U>(
    x: &ArrayD<T>,
    counts: &ArrayD<U>,
    eps: T,
    threshold: T,
    mark: i64,
) -> (ArrayD<T>, ArrayD<U>)
where
    T: Float + NumCast + ToPrimitive + Debug + Clone + ScalarOperand + NdFloat,
    U: Clone + ToPrimitive + Zero + Debug + num_traits::PrimInt + ScalarOperand,
{
    let x_shape = x.shape();
    let (batch_size, nodes, mems) = (x_shape[0], x_shape[1], x_shape[2]);
    let normal_path = argmaxi(x, eps);
    let condition = x.mapv(|value| if value > threshold { 1 } else { 0 });
    let condition_dim = condition.ndim();
    let trigger_growth = sum_generic(&condition, Axis(condition_dim - 1), true);
    let trigger_growth = trigger_growth.mapv(|value| value <= 0);
    let (avail, all) = mark_reserved_indices(&normal_path, &counts, &trigger_growth, mark);
    let avail = move_to_back(&avail, mark);
    let avail = expand_for_batches(&avail, batch_size).unwrap();
    let all = expand_for_batches(&all, batch_size).unwrap();

    let avail_is_mark = avail.mapv(|value| value == mark);
    let finalized = where_op(&avail_is_mark, &all, &avail).unwrap();
    let indices_sg = transpose_dims(&finalized, finalized.ndim() - 1, finalized.ndim() - 2);
    let reshape_shape: Vec<i64> = vec![batch_size as i64, -1, 1];
    let indices_sg = reshape(&indices_sg, &reshape_shape).unwrap();

    let mut growth_path = zeros_like(&normal_path);
    let values_of_interest = gather(&x, (x.ndim() - 1) as isize, &indices_sg).unwrap();
    let values_of_interest_eq_0 = values_of_interest.mapv(|value| value == T::zero());
    let values_of_interest = where_value(
        &values_of_interest_eq_0,
        &values_of_interest,
        NumCast::from(-1).unwrap(),
    )
    .unwrap();
    let scatter_dim = growth_path.ndim() - 1;
    let _ = scatter(
        &mut growth_path,
        scatter_dim as isize,
        &indices_sg,
        &values_of_interest,
    );
    let growth_path = abs_ndarray(&growth_path);
    let growth_path = argmaxi(&growth_path, eps);
    let trigger_growth = expand_at_dim(
        &trigger_growth,
        trigger_growth.ndim() - 1,
        normal_path.shape()[2],
    )
    .unwrap();
    let grown = where_op(&trigger_growth, &growth_path, &normal_path).unwrap();

    // let updated_counts = reduce(&grown, "bnm,bnm->nm").unwrap();
    let updated_counts = reduce_specialized::reduce_bnm_to_nm(&grown.view()).unwrap();
    let updated_counts_leq_0 = updated_counts.mapv(|value| value <= T::zero());
    let updated_counts = updated_counts.mapv(|value| NumCast::from(value).unwrap_or(U::zero()));
    let updated_counts = where_op(&updated_counts_leq_0, &counts, &updated_counts).unwrap();

    // return both grown and updated_counts
    (grown, updated_counts)
}

File: reference/growth_argmaxi.py
import torch
from util.save_reference import save_reference
from typing import Tuple, List, Iterable
from maxi import create_tensors as create_tensors_maxi
from argmaxi import create_tensors as create_tensors_argmaxi
from mark_reserved_indices import create_tensors as create_tensors_mark_reserved_indices
from move_value_to_back import create_tensors as create_tensors_move_value_to_back
from expand_for_batches import create_tensors as create_tensors_expand_for_batches
import einops

def create_tensors(
        shape: Tuple[int, ...] | List[int] | Iterable[int],
        eps = 1e-8,
        threshold = 0.5,
        dtype: torch.dtype = torch.float32,
        dir: str = "data",
        name: str = "growth_argmaxi",
        x = None,
        counts = None
):
    assert len(shape) == 3, "Shape must be of length 3"
    x = torch.rand(shape, dtype=dtype) if x is None else x
    counts = torch.randint(0, 4, shape[1:], dtype=torch.float32) if counts is None else counts
    
    batch_size, nodes, mems = x.shape

    _, normal_path = create_tensors_argmaxi(shape=None, x=x)
    trigger_growth = torch.sum(x > threshold, dim=-1, keepdim=True) <= 0
    mark = -2
    avail, all, _, _, _ = create_tensors_mark_reserved_indices(shape=None, mark=mark, acts=normal_path, counts=counts, trigger=trigger_growth)
    avail, _ = create_tensors_move_value_to_back(shape=None, x=avail)
    avail, _ = create_tensors_expand_for_batches(shape=None, x=avail, batch_size=batch_size)
    all, _ = create_tensors_expand_for_batches(shape=None, x=all, batch_size=batch_size)
    final = torch.where(avail == mark, all, avail)
    indices_sg = final.transpose(-1, -2).reshape(batch_size, -1, 1)
    growth_path = torch.zeros_like(normal_path)
    values_of_interest = torch.gather(x, -1, indices_sg)
    values_of_interest[values_of_interest == 0] = 1
    growth_path = torch.scatter(growth_path, -1, indices_sg, values_of_interest)
    growth_path = torch.abs(growth_path)
    _, growth_path = create_tensors_argmaxi(shape=None, x=growth_path)
    grown = torch.where(trigger_growth, growth_path, normal_path)

    updated_counts = einops.reduce(grown, 'batch nodes mems -> nodes mems', 'sum')
    updated_counts = torch.where(updated_counts <= 0, counts, updated_counts)

    return grown, updated_counts, x, counts

def create_growth_argmaxi(
        shape: Tuple[int, ...] | List[int] | Iterable[int],
        eps = 1e-8,
        threshold = 0.5,
        dtype: torch.dtype = torch.float32,
        dir: str = "data",
        name: str = "growth_argmaxi",
):
    grown, updated_counts, x, counts = create_tensors(shape, eps, threshold, dtype, dir, name)
    save_reference(grown, dir, f"{name}_grown")
    save_reference(updated_counts, dir, f"{name}_updated_counts")
    save_reference(x, dir, f"{name}_x")
    save_reference(counts, dir, f"{name}_counts")

if __name__ == "__main__":
    create_growth_argmaxi((6, 7, 8), eps=1e-8, threshold=0.5, dtype=torch.float32, dir="data", name="growth_argmaxi")